sbioapputils/app_runner/workflow_utils.py

```diff
@@ -0,0 +1,123 @@
+import yaml
+
+
+def _parse_workflow(request):
+    """Helper function to parse the workflow configuration."""
+    if request.get('workflow_name'):
+        workflow_loc = f"app/{request['workflow_name']}.yml"
+    else:
+        workflow_loc = "app/workflow.yml"
+        
+    with open(workflow_loc, "r") as stream:
+        try:
+            yaml_dict = yaml.safe_load(stream)
+        except yaml.YAMLError as exc:
+            print(exc)
+            
+    name = yaml_dict['name']
+    stages = yaml_dict['stages']
+    parameters = yaml_dict['parameters']
+    return name, stages, parameters
+
+
+def dir_path(string):
+    """Function to check if the specified string is a directory."""
+    if os.path.isdir(string):
+        return string
+    else:
+        raise NotADirectoryError(string)
+
+
+def validate_config(request, job_id):
+    """Function to validate the request config."""
+    name, stages, parameters = _parse_workflow(request)
+    for key, value in request['input_files'].items():
+        request[key] = value
+    request['job_id'] = job_id
+
+    # Check all required parameters are provided
+    no_default = []
+    no_type = []
+    wrong_data_types = []
+    invalid_value = []
+
+    for key in parameters.keys():
+
+        # Check if default is present
+        if not parameters[key].get('default'):
+            no_default.append(key)
+        elif key not in request:
+            request[key] = parameters[key]['default']
+
+        # Check if type is present
+        if not parameters[key].get('type'):
+            no_type.append(key)
+        elif not isinstance(request[key], parameters[key]['type']):
+            wrong_data_types.append(key)
+
+        if parameters[key].get('user_defined') == 'True':
+            if parameters[key]['type'] in ['int', 'float']:
+                # Check between min and max
+                if parameters[key].get('max_value'):
+                    if request[key] > parameters[key]['max_value']:
+                        invalid_value.append(key)
+                if parameters[key].get('min_value'):
+                    if request[key] < parameters[key]['min_value']:
+                        invalid_value.append(key)
+
+            elif parameters[key]['type'] == 'str':
+                dropdown = not parameters[key].get('from_data') == 'True'
+                # Category settings
+                if dropdown:
+                    if request[key] not in parameters[key]['options']:
+                        invalid_value.append(key)
+
+        # Create directory if needed
+        try:
+            if str(request[key])[-1] == '/':
+                if not os.path.exists(request[key]):
+                    os.mkdir(request[key])
+        except KeyError:
+            pass
+
+    if not all(param in request.keys() for param in no_default):
+        raise Exception(f"These required parameters are not specified: {no_default}")
+
+    if not all(param in request.keys() for param in no_type):
```

Comment:
what is this code checking?

Reply:
If the workflow yaml file does not have 'default' specified, and the request has not provided a value either. i.e. we do not know what value a required parameter should have